### PR TITLE
WIP Add a check that flows are correctly stored on the server

### PR DESCRIPTION
**Pull request by mfeurer**
by checking if they're equal after downloading them again in the publish function. Fixes issue #145.

### CODE REVIEW

File: openml/flows/flow.py [hunk 1]
```diff
@@ -338,9 +338,24 @@ def publish(self):
...
         return self
 
 
 def _add_if_nonempty(dic, key, value):
     if value is not None:
         dic[key] = value
+
+
+def _check_flow(flow):
+    import openml.flows.functions
```
Comment (janvanrijn) [inline openml/flows/flow.py]: Imports at the top please!

File: openml/flows/functions.py [hunk 2]
```diff
@@ -127,4 +127,47 @@ def _list_flows(api_call):
...
+    if not isinstance(flow1, OpenMLFlow):
+        raise TypeError('Argument 1 must be of type OpenMLFlow, but is %s' %
+                        type(flow1))
+
+    if not isinstance(flow2, OpenMLFlow):
+        raise TypeError('Argument 2 must be of type OpenMLFlow, but is %s' %
+                        type(flow2))
+
+    # Name is actually not generated by the server, but it will be
+    # tested further down with a getter (allows mocking in the tests)
```
Comment (janvanrijn) [inline openml/flows/functions.py]: I don't understand this comment. 'Name' is not in the array, right?

File: openml/flows/functions.py [hunk 3]
```diff
@@ -127,4 +127,47 @@ def _list_flows(api_call):
...
+                if not name in attr2:
+                    raise ValueError('Component %s only available in '
+                                     'argument2, but not in argument1.' % name)
+                assert_flows_equal(attr1[name], attr2[name])
+
+        else:
+            if attr1 != attr2:
+                raise ValueError("Flow %s: values for attribute '%s' differ: "
+                                 "'%s' vs '%s'." %
+                                 (str(flow1.name), str(key), str(attr1), str(attr2)))
```
Comment (janvanrijn) [inline openml/flows/functions.py]: I am very curious whether this will work for parameter dicts ...

File: openml/flows/flow.py (hunk 1 above)
Comment (mfeurer) [inline openml/flows/flow.py]: Importing at the top won't work here due to cyclic imports. I'll add a comment mentioning this.

File: openml/flows/functions.py (hunk 2 above)
Comment (mfeurer) [inline openml/flows/functions.py]: Thanks for spotting this. The comment is outdated and I'll remove it.

File: openml/flows/functions.py (hunk 3 above)
Comment (mfeurer) [inline openml/flows/functions.py]: Yes, that's why the unit tests still work ;)

**Pull request merged**